tools/debugstream/thread-info.py

```diff
@@ -0,0 +1,217 @@
+	chunk = None
+	core_data = []
+
+	def set_file(self, f):
+		self.f = f
+
+	def decode_chunk(self):
+		if self.chunk != None:
+			return
+		self.f.seek(0)
+		slot = self.f.read(self.file_size)
+		if not get_telemetry2_payload_ok(slot):
+			return
+		self.chunk_offset = get_telemetry2_chunk_offset(TELEMETRY2_ID_THREAD_INFO, slot)
+		if self.chunk_offset < 0:
+			print("Thread info chunk not found")
+			self.chunk = None
+			return
+		self.chunk = ctypes.cast(slot[self.chunk_offset:],
+					 ctypes.POINTER(ThreadInfoChunk))
+		hdr = self.chunk.contents.hdr
+		#print(f"Chunk at {self.chunk_offset} id: {hdr.id} size: {hdr.size}")
+		print(f"core_count: {self.chunk.contents.core_count}")
+
+	def init_core_data(self):
+		if self.chunk == None:
+			return
+		if len(self.core_data) == self.chunk.contents.core_count:
+			return
+		self.core_data = [None] * self.chunk.contents.core_count
+		for i in range(self.chunk.contents.core_count):
+			offset = self.chunk_offset + self.chunk.contents.core_offsets[i]
+			self.core_data[i] = CoreData(offset, i)
+			#print(f"core_offsets {i}: {self.chunk.contents.core_offsets[i]}")
+
+	def loop_cores(self):
+		if len(self.core_data) == 0:
+			return
+		self.f.seek(0)
+		slot = self.f.read(self.file_size)
+		for i in range(len(self.core_data)):
+			self.core_data[i].print(slot)
+
+	def reset(self):
+		self.f = None
+		self.chunk = None
+
+	def print_status(self, tag):
+		print(f"{tag} f {self.f} offset {self.chunk_offset} chunk {self.chunk} cores {len(self.core_data)}")
+
+def main_f(args):
+	"""
+	Map telemetry2 slot into memorey and find the chunk we are interested in
+	"""
+	decoder = Telemetry2ThreadAnalyzerDecoder()
+	prev_error = None
+	while True:
+		try:
+			with open(args.telemetry2_file, "rb") as f:
+				decoder.set_file(f)
+				decoder.decode_chunk()
+				decoder.init_core_data()
+				while True:
+					decoder.loop_cores()
+					time.sleep(args.update_interval)
+		except FileNotFoundError:
+			print(f"File {args.telemetry2_file} not found!")
+			break
+		except OSError as e:
+			if str(e) != prev_error:
+				print(f"Open {args.telemetry2_file} failed '{e}'")
+				prev_error = str(e)
+			decoder.reset()
+		time.sleep(args.update_interval)
+
+def parse_params():
+	""" Parses parameters
+	"""
+	parser = argparse.ArgumentParser(description=
+					 "SOF Telemetry2 thread info client. "+
+					 "Opens telemetry2 slot debugfs file, looks"+
+					 " for thread info chunk, and decodes its "+
+					 "contentes into stdout periodically.")
+	parser.add_argument('-t', '--update-interval', type=float,
+			    help='Telemetry2 window polling interval in seconds, default 1',
+                            default=1)
+	parser.add_argument('-f', '--telemetry2-file',
+			    help='File to read the telemetry2 data from, default /sys/kernel/debug/sof/telemetry2',
+                            default="/sys/kernel/debug/sof/telemetry2")
+	parsed_args = parser.parse_args()
+	return parsed_args
+
+args = parse_params()
+main_f(args)
```

Comment:
this is supposed to run on the target system, right? So cross-OS compatibility isn't needed? So there should be no objections against having it in C / C++ ;-)

Reply:
I was explicitly advised to use python, I am not going to change it to C now.